### DISCUSSION TITLE
Beginnings of a new README file

### DISCUSSION THREAD

**Discussion by blakemcbride** [modula3/cm3#1192] [General]
Greetings,

I have begun work on a new README.md file intened to be added to the repo at the top-level  It's at [https://blakemcbride.us/files/README.md](https://blakemcbride.us/files/README.md)

Your feedback is requested.

Questions.

1. Is the x86, integrate back-end, 32 or 64 bits?

2. Once the system is bootstrapped with the C back-end, how do you build for the GNU back-end?  Do you need the GNU compiler source?  Do you need a specific version of the GNU compiler?

Again, I think it is very important that all of the information on how to build the system be a part of the distribution file.

Thanks!

Blake

**Comment by blakemcbride** [comment]
I just made an update to the file.

**Comment by RodneyBates** [comment]
On 11/20/24 06:54, Blake McBride wrote:
>
> Greetings,
>
> I have begun work on a new README.md file intened to be added to the repo at the top-level It's at https://blakemcbride.us/files/README.md
>
> Your feedback is requested.
>

Good start.

Unless something has changed recently( Jay?) the C back end produces C, which
is compiled by your favorite C compiler.  But the link step for this
uses A C++ compiler as front to a native linker.

The x86 backend produces code for at least either Windows or Linux, maybe more.
In fact, I believe it produces assembly code and then runs a native assembler.


> Questions.
>
> 1.
>
>     Is the x86, integrate back-end, 32 or 64 bits?
>
It will do either

> 1.
>
>
> 2.
>
>     Once the system is bootstrapped with the C back-end, how do you build for the GNU back-end? Do you need the GNU compiler source? Do you need a specific version of the GNU compiler?
>

Can't answer how off the top of my head.  Others here likely can. If nobody responds soon, I'll
dredge out the process.

The GCC backend source is all in the Modula-3 distribution.  I guess you'd call it a fork.  Irrelevant
stuff is pruned out and what might be called a binding is added that reads the front-end's
output from a file and builds GCC's early intermediate data structure.  No need for regular
GCC.

In contrast, the LLVM backend uses an unmodified LLVM distribution, of a certain release.
(Peter?).  It contains a Modula-3 binding to LLVM's data structure building calls.

**Reply by blakemcbride** [reply]
I am confused regarding C vs. C++. I see both .c and .cpp files in the bootstrap directory of the distribution. I could use some clarification.

Details on the native back end would be helpful.

Using a fork of GCC scares me. First, you lose the benefit of compiler corrections. Second, you lose the ability to support new architectures. In addition to the immense complexity, these are the exact reasons I don't like the GCC back end. Same issue with "certain release" of LLVM.  Same for the native back end.

It is my opinion that doing anything except the C back end is putting a timestamp on the life of CM3.

**Reply by VictorMiasnikov** [reply]
> doing anything except ( . . .) 
  
  We should first solve https://github.com/modula3/cm3/discussions/1188#discussioncomment-12188280
  
  P.S.
  
"Modula-3 binding to LLVM" on AMD64_LINUX don't affected. I.e. it is good as "m3сс backend"
